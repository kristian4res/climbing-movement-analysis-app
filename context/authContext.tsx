import React from 'react';
import { useStorageState } from '../hooks/useStorageState';
import { router, useNavigation } from 'expo-router';

const AuthContext = React.createContext<{
    signIn: () => void;
    signOut: () => void;
    session?: string | null;
    isLoading: boolean;
}>({
    signIn: () => null,
    signOut: () => null,
    session: null,
    isLoading: false,
});

export function useSession() {
    const value = React.useContext(AuthContext);
    if (process.env.NODE_ENV !== 'production') {
        if (!value) {
            throw new Error('useSession must be wrapped in a <SessionProvider />');
        }
    }

    return value;
}

export function SessionProvider(props: React.PropsWithChildren) {
  const [[isLoading, session], setSession] = useStorageState('session');
  const navigator = useNavigation();

  return (
    <AuthContext.Provider
      value={{
        signIn: () => {
          setSession('example');
          console.log('Signed in user: ', 'example');
          router.replace('/(tabs)');
        },
        signOut: () => {
          setSession(null);
        },
        session,
        isLoading,
      }}>
      {props.children}
    </AuthContext.Provider>
  );
}
