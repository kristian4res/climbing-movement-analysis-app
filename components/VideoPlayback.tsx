import React, { useState, useEffect, useContext } from 'react';
import { Video, ResizeMode } from 'expo-av';
import { Alert, View, StyleSheet } from 'react-native';
import { Text, Button } from 'react-native-paper';
import { router } from 'expo-router';
import * as MediaLibrary from 'expo-media-library';
import { SCREEN_WIDTH } from '@/constants/Dimensions';
import { ThemeContext } from '@/context/themeContext';
import Colors from '@/constants/Colors';

export default function VideoPlayback({ children, filename }: { children?: React.JSX.Element, filename: string }) {
    const { isDark } = useContext(ThemeContext);
    const containerStyle = isDark ? Colors.darkContainer : Colors.lightContainer;
    const [videoLoaded, setVideoLoaded] = useState(false);
    const [videoUri, setVideoUri] = useState<string | null>(null);

    const fetchVideoUri = async () => {
        if (!filename) {
            console.log('No filename provided');
            return;
        }
        
        const album = await MediaLibrary.getAlbumAsync('cma-app');
        if (!album) {
            console.log('Album not found');
            return;
        }
        
        const { assets } = await MediaLibrary.getAssetsAsync({
            mediaType: 'video',
            album: album,
        });
        const videoAsset = assets.find(asset => {
            return asset.filename === filename
        });

        if (videoAsset) {
            setVideoUri(videoAsset.uri);
            setVideoLoaded(true);
        } else {
            console.log('Could not find video: ', filename);
            return;
        }
    };

    useEffect(() => {
        fetchVideoUri();
    }, [filename]);

    return (
        <View style={{...containerStyle, justifyContent: 'flex-start'}}>
            <Text style={styles.filename}>{filename}</Text>
            {videoLoaded ? (
                <View style={styles.videoContainer}>
                    <Video
                        source={{ uri: videoUri ? videoUri : '' }}
                        rate={1.0}
                        volume={1.0}
                        isMuted={false}
                        resizeMode={ResizeMode.COVER}
                        isLooping={false}
                        useNativeControls
                        style={styles.video}
                        onError={(e) => console.log('Error with displaying video: ', e)} 
                    />
                    {
                        children ? children 
                        : 
                        <Button
                            onPress={() => {
                                if (!videoUri) {
                                    console.log('No video URI available');
                                    return;
                                }
                                Alert.alert(
                                    'Delete Video',
                                    'Are you sure you want to delete this video?',
                                    [
                                        {
                                            text: 'Cancel',
                                            style: 'cancel',
                                        },
                                        {
                                            text: 'OK',
                                            onPress: async () => {
                                                try {
                                                    const album = await MediaLibrary.getAlbumAsync('cma-app');
                                                    if (!album) {
                                                        console.log('Album not found');
                                                        return;
                                                    }
                                                    const assets = await MediaLibrary.getAssetsAsync({
                                                        album: album,
                                                        first: 1,
                                                        mediaType: 'video',
                                                        sortBy: 'creationTime',
                                                    });
                                                    const asset = assets.assets.find(a => a.uri === videoUri);
                                                    if (!asset) {
                                                        console.log('Asset not found');
                                                        return;
                                                    }
                                                    console.log('Deleting video: ', asset);
                                                    await MediaLibrary.deleteAssetsAsync([asset.id]);
                                                    console.log('Deleted video');
                                                    router.replace('/(tabs)/climbs');
                                                } catch (error) {
                                                    console.log('Error: ', error)
                                                }
                                            },
                                        },
                                    ],
                                    { cancelable: false },
                                );
                            }}
                        >
                            Delete Video
                        </Button>
                    }
                </View>
            ) : (
                <Text style={styles.noVideoText}>No video to display</Text>
            )}
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'flex-start',
        alignItems: 'center',
    },
    filename: {
        color: 'white',
        fontSize: 16,
    },
    videoContainer: {
        flex: 1,
    },
    video: {
        flex: 0.9,
        width: SCREEN_WIDTH,
    },
    noVideoText: {
        color: 'white',
        fontSize: 16,
    },
});