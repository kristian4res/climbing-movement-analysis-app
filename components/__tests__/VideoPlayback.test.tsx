import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeContext } from '@/context/themeContext';
import VideoPlayback from '@/components/VideoPlayback';

jest.mock('expo-media-library', () => ({
    getAlbumAsync: jest.fn(() => Promise.resolve({})),
    getAssetsAsync: jest.fn(() => Promise.resolve({ assets: [] })),
    deleteAssetsAsync: jest.fn(),
}));

describe('VideoPlayback component', () => {
    it('matches the snapshot', () => {
        const testFilename = 'Test filename';

        const { toJSON } = render(
            <ThemeContext.Provider value={{ isDark: false, toggleTheme: () => {} }}>
                <VideoPlayback filename={testFilename} />
            </ThemeContext.Provider>
        );

        expect(toJSON()).toMatchSnapshot();
    });

    it('renders the passed filename', async () => {
        const testFilename = 'Test filename';

        const { findByText } = render(
            <ThemeContext.Provider value={{ isDark: false, toggleTheme: () => {} }}>
                <VideoPlayback filename={testFilename} />
            </ThemeContext.Provider>
        );

        expect(await findByText(testFilename)).toBeTruthy();
    });
});