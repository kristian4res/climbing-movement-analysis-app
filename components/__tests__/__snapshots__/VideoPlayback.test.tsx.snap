// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayback component matches the snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "rgb(250, 253, 251)",
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "fontFamily": "System",
          "fontWeight": "400",
          "letterSpacing": 0,
        },
        {
          "writingDirection": "ltr",
        },
        {
          "color": "white",
          "fontSize": 16,
        },
      ]
    }
  >
    Test filename
  </Text>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "fontFamily": "System",
          "fontWeight": "400",
          "letterSpacing": 0,
        },
        {
          "writingDirection": "ltr",
        },
        {
          "color": "white",
          "fontSize": 16,
        },
      ]
    }
  >
    No video to display
  </Text>
</View>
`;
